fix(api, shared-data): Allow tough universal lid to stack universally

CaseyBatten opened the pull request:
# Overview
Covers RABR-777

Update the way we validate stacking of labware to ensure that a labware which utilizes the "default" stacking offset flag and DOES NOT include a compatible parent list can stack on any labware.

Universal Lid Stacking offset of 0.8mm based on overall height of 8.8mm - stacked measured height of 8.0mm:
![image (1)](https://github.com/user-attachments/assets/15574a30-ad65-4698-81b5-bb644a203993)



## Test Plan and Hands on Testing
- [x] The following protocol should pass analysis and run on a robot:
```
from opentrons.protocol_api import ProtocolContext

metadata = {
    "protocolName": "Tip rack basic labware stuff",
    "author": "Opentrons <protocols@opentrons.com>",
}
requirements = {
    "robotType": "Flex",
    "apiLevel": "2.23",
}

def run(protocol: ProtocolContext) -> None:
    trash = protocol.load_trash_bin("A3")
    universal_stack = protocol.load_lid_stack(load_name="opentrons_tough_universal_lid", location="A1", quantity=2)
    plate = protocol.load_labware("nest_96_wellplate_200ul_flat", "B2")
    plate2 = protocol.load_labware("nest_96_wellplate_200ul_flat", "B3", lid="opentrons_tough_universal_lid")
    protocol.move_lid(plate2, trash, True)
    protocol.move_lid(universal_stack, plate, True)
    protocol.move_lid(universal_stack, plate2, True)
```

## Changelog
- Updated the parent labware validation to pass for defaults
- Update the universal lid definition to give it stacking offsets for the lid stack object, opentrons well plate, and themselves

## Review requests
Is the risk assessment below a big red flag?

## Risk assessment
Medium to low. Currently this means that new labware made (and this one) with a default stacking offset and no compatible parents can go on ANYTHING. This means we can use it on tipracks, and tube racks, and all the horrible shapes therein.

CaseyBatten committed:
allow labware with default stacking offset but no compatible parents to stack on everything

CaseyBatten commented:
> This is just for the Stacker, right? I assume it does not need to be included in RS 8.5.0?

No this isn't related to stacker. This is for the tough universal lid labware. It is meant to be able to stack on top of "anything" (for now). In the ticket that created this issue we listed that this labware needed it's `compatibleParentLabware` list added and extended, but this list would in theory be quite large and continuously updated into the future as more labware it is cross compatible with are added. 

As for whether or not this needs to go on 8.5.0, I think for now we're safe targeting this to edge as the labware won't be released yet anyways, and if we want we can cherry pick back to 8.5.

CaseyBatten committed:
enforce that labware with a default stacking offset must have no compatible parents in order to be universal

SyntaxColoring commented on api/src/opentrons/protocol_engine/resources/labware_validation.py:
```diff
@@ -48,7 +48,13 @@ def validate_labware_can_be_stacked(
 ) -> bool:
     """Validate that the labware being loaded onto is in the above labware's stackingOffsetWithLabware definition."""
-    return below_labware_load_name in top_labware_definition.stackingOffsetWithLabware
+    return (
+        below_labware_load_name in top_labware_definition.stackingOffsetWithLabware
+        or (
+            "default" in top_labware_definition.stackingOffsetWithLabware
+            and top_labware_definition.compatibleParentLabware is None
+        )
+    )
```
This snippet of logic looks right, but our logic overall for "can this be stacked, and if so, how" is getting pretty scattered. For example, it's not clear to me what the distinction is between `LabwareView.raise_if_labware_cannot_be_stacked()` and `labware_validation.validate_labware_can_be_stacked()`. It seems like only one of those ought to exist?

And then there's this logic directly in `LoadLabwareImplementation`, which seems like it should live in `LabwareView.raise_if_labware_cannot_be_stacked()` or 
`labware_validation.validate_labware_can_be_stacked()`?

https://github.com/Opentrons/opentrons/blob/a56a55c164ed8075ea46134727ef8688d58b9227/api/src/opentrons/protocol_engine/commands/load_labware.py#L168-L186

Can we centralize and deduplicate?

SyntaxColoring commented on shared-data/labware/definitions/2/opentrons_tough_universal_lid/1.json:
```diff
@@ -47,9 +47,24 @@
+      "y": 0,
+      "z": 0
+    },
+    "opentrons_tough_universal_lid": {
+      "x": 0,
+      "y": 0,
+      "z": 0
```
Is (0, 0, 0) really correct when this lid is stacked atop itself? The very bottom of the top instance's bounding box aligns exactly with the very top of the bottom instance's bounding box, without any overlap?

SyntaxColoring commented on shared-data/labware/definitions/2/opentrons_tough_universal_lid/1.json:
```diff
@@ -47,9 +47,24 @@
       "y": 0,
       "z": 6.6
+    },
+    "opentrons_96_wellplate_200ul_pcr_full_skirt": {
+      "x": 0,
+      "y": 0,
+      "z": 0
+    },
```
Is (0, 0, 0) really correct when this lid is stacked atop this well plate? The very bottom of the lid's bounding box aligns exactly with the very top of the well plate's bounding box, without any overlap?

SyntaxColoring reviewed (changes requested):
This fundamentally seems right to me, just some weedy code-level comments. Changes requested for the code duplication thing, but let me know if that's too hard to do here.

> Is the risk assessment below a big red flag?
>
> ...Currently this means that new labware made (and this one) with a default stacking offset and no compatible parents can go on ANYTHING. This means we can use it on tipracks, and tube racks, and all the horrible shapes therein.

Yeah. If we have to choose between allowing stacks that don't make sense, or having to enumerate every possible stack in the labware definitions, I definitely think allowing stacks that don't make sense is the lesser evil.

CaseyBatten committed:
correction to ensure no regression for old lid stacking

CaseyBatten committed:
addition of compatible parents to old version 1 file

github-actions[bot] committed:
fix(analyses-snapshot-testing): heal RABR_777_universal_stacking_fix snapshots (#18467)

This PR was requested on the PR
https://github.com/Opentrons/opentrons/pull/18460

Co-authored-by: CaseyBatten <48519493+CaseyBatten@users.noreply.github.com>

sfoster1 commented on shared-data/labware/definitions/2/opentrons_tough_universal_lid/1.json:
```diff
@@ -47,9 +47,24 @@
     }
   },
-  "stackLimit": 1,
+  "stackLimit": 2,
```
let's just raise this a bunch tbh. make it like 5 or something. we should get rid of it at some point

CaseyBatten replied:
I'll make it 5, the limit was set to 2 in Rhyann's original change which is where this current number came from. I do think we should keep the stack limit as a concept. It definitely plays a real role for the auto-sealing lids, which become unstable when stacked more than 5 high.

CaseyBatten committed:
definition adjustments and labware raise case consolidation

sfoster1 reviewed (approved):
Looks good to me, though I agree with Max that we should centralize some logic at this point.

sfoster1 reviewed (changes requested):
Actually, putting a hold on this to discuss 8.5 inclusion - @ddcc4 we should probably have this in 8.5

sfoster1 reviewed (approved):
This looks good to me!

SyntaxColoring commented on shared-data/labware/definitions/2/opentrons_tough_universal_lid/1.json:
```diff
@@ -47,9 +47,24 @@
       "x": 0,
       "y": 0,
       "z": 6.6
+    },
+    "opentrons_96_wellplate_200ul_pcr_full_skirt": {
+      "x": 0,
+      "y": 0,
+      "z": 7.1
+    },
+    "protocol_engine_lid_stack_object": {
+      "x": 0,
+      "y": 0,
+      "z": 0
+    },
+    "opentrons_tough_universal_lid": {
+      "x": 0,
+      "y": 0,
+      "z": 0.8
```
Does the `default` offset intentionally not match any of these offsets? I think for the other lid, `default` was chosen to match `opentrons_96_wellplate_200ul_pcr_full_skirt`, right?

CaseyBatten replied:
So on the diff between the `validate` and `raise` function sets here, I've consolidated our only other stacking error into our raise logic, but I left validate with minimum changes. This function is called quite a few places, and we'd have unintended knock on effects if we did things like enforcing the compatible parent list against the stacking offsets there. One such knock on effect would be breaking labware compatibility with adapters. We could add logic to pass in those cases, but I'm not confident we wouldn't miss other weird combinations we currently allow that aren't described in our labware definitions as explicitly as the parent/lid relationships are. I do think we should clean it up and combine them, but its a bit too much for just this lid labware PR.

Regarding the default value, that came from the original labware definition. Micco is identifying some specific overlaps for the planned supported labware, so we may change the default to a more common case like the opentrons 96 well plate in the future. For now it hasn't been decided what we want that final value to be.

SyntaxColoring reviewed (approved):
Thanks!

One comment double-checking the offsets in the lid definition.

Also still kind of confused by the split between `raise_if_labware_cannot_be_stacked()` and `validate_labware_can_be_stacked()`, but for the sake of unblocking people, that can be resolved separately.

CaseyBatten committed:
Merge branch 'chore_release-8.5.0' into RABR_777_universal_stacking_fix
